{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "dbc3af35",
   "metadata": {},
   "source": [
    "# Gumshoe Tutorial\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98ccd0e1",
   "metadata": {},
   "source": [
    "## Introduction\n",
    "This tutorial is based on aggregated single-cell data from the following paper:\n",
    "- Usoskin D, Furlan A, Islam S, Abdo H, Lönnerberg P, Lou D, Hjerling-Leffler J, Haeggström J, Kharchenko O, Kharchenko PV, Linnarsson S, Ernfors P (2015) Unbiased classification of sensory neuron types by large-scale single-cell RNA sequencing. Nat Neurosci 18:145–153.\n",
    "\n",
    "For the sake of simplicity and as the focus of Gumshoe is supplemental utilities for Sleuth, we can skip the step of transcript abundance quantification using Kallisto. We have done this step, and the output files are below. A corresponding metadata file for the dataset has been included. The data is from a single-cell experiment on colonic sensory neurons in mice. The data has been combined and converted into bulk RNA-Seq data, and has four tissue sources (NF, NP, PEP, TH) and two sexes (male and female)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79993a01",
   "metadata": {},
   "source": [
    "## Downloading The Data\n",
    "Tutorial data can be found in the data folder of the package. For this tutorial, you should find the following in the folder:\n",
    "```Bash\n",
    "Tutorial_Gumeshoe_Functions\n",
    "├── gumshoe_functions_analysis.R\n",
    "├── gumshoe_functions_tutorial.ipynb\n",
    "├── metadata.txt\n",
    "├── output\n",
    "│   ├── female_NF.0.kallisto\n",
    "│   ├── female_NF.1.kallisto\n",
    "│   ├── female_NF.2.kallisto\n",
    "│   ├── female_NP.0.kallisto\n",
    "│   ├── female_NP.1.kallisto\n",
    "│   ├── female_NP.2.kallisto\n",
    "│   ├── female_PEP.0.kallisto\n",
    "│   ├── female_PEP.1.kallisto\n",
    "│   ├── female_PEP.2.kallisto\n",
    "│   ├── female_TH.0.kallisto\n",
    "│   ├── female_TH.1.kallisto\n",
    "│   ├── female_TH.2.kallisto\n",
    "│   ├── male_NF.0.kallisto\n",
    "│   ├── male_NF.1.kallisto\n",
    "│   ├── male_NF.2.kallisto\n",
    "│   ├── male_NP.0.kallisto\n",
    "│   ├── male_NP.1.kallisto\n",
    "│   ├── male_NP.2.kallisto\n",
    "│   ├── male_PEP.0.kallisto\n",
    "│   ├── male_PEP.1.kallisto\n",
    "│   ├── male_PEP.2.kallisto\n",
    "│   ├── male_TH.0.kallisto\n",
    "│   ├── male_TH.1.kallisto\n",
    "│   └── male_TH.2.kallisto\n",
    "├── plot\n",
    "│   ├── pca_plot.pdf\n",
    "│   └── sample_volc_plot.jpeg\n",
    "└── results\n",
    "    ├── NF Metadata\n",
    "    └── TH Metadata\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1c4206d",
   "metadata": {},
   "source": [
    "---\n",
    "## Analysis Using Sleuth and Gumshoe \n",
    "\n",
    "***NOTE: Please ensure that your working directory is the current directory, as this is critical for the remainder of the tutorial.***\n",
    "\n",
    "### Loading Libraries\n",
    "After downloading or generating the data, our first step should be loading the libraries we will be using throughout the program.\n",
    "\n",
    "```R\n",
    "library(tidyverse)\n",
    "library(sleuth)\n",
    "library(biomaRt)\n",
    "library(gumshoe)\n",
    "library(ggrepel)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7ad1794",
   "metadata": {},
   "source": [
    "### Metadata File Creation\n",
    "Sleuth requires a metadata file, and, in our case, the file contains the sample name, path, sex, and tissue type. If you are well-versed with RNASeq analysis, feel free to create this file yourself and load it in; otherwise, we can either use the metadata.txt file that is provided alongside the downloadable data or run the following code to create our own:\n",
    "\n",
    "```R\n",
    "# Create a dataframe that is 4 columns by 24 rows and rename the columns\n",
    "metadata <- data.frame(matrix(NA, nrow = 24, ncol = 4))\n",
    "colnames(metadata) <- c(\"sample\", \"path\", \"sex\", \"tissue\")\n",
    "\n",
    "# Assign sample names\n",
    "# The naming format is sex, tissue type, and replicate number\n",
    "metadata[1] <- c(\"female_NF.0\", \"female_NF.1\", \"female_NF.2\", \"female_NP.0\", \"female_NP.1\", \"female_NP.2\", \"female_PEP.0\", \"female_PEP.1\", \"female_PEP.2\", \"female_TH.0\", \"female_TH.1\", \"female_TH.2\", \"male_NF.0\", \"male_NF.1\", \"male_NF.2\", \"male_NP.0\", \"male_NP.1\", \"male_NP.2\", \"male_PEP.0\", \"male_PEP.1\", \"male_PEP.2\", \"male_TH.0\", \"male_TH.1\", \"male_TH.2\")\n",
    "\n",
    "# Assign the file path\n",
    "# NOTE: This line of code assumes that your data is contained in the working directory in a subfolder named \"output\"\n",
    "for (cell in 1:24){\n",
    "metadata[cell, 2] <- paste(\"output/\", metadata[cell, 1], sep = \"\")\n",
    "metadata[cell, 2] <- paste(metadata[cell, 2], \".kallisto\", sep = \"\")\n",
    "}\n",
    "\n",
    "# Assign sex\n",
    "metadata[1:12, 3] <- \"F\"\n",
    "metadata[13:24, 3] <- \"M\"\n",
    "\n",
    "# Assign the tissue type\n",
    "for (cell in 1:24){\n",
    "metadata[cell, 4] <- unlist(strsplit(metadata[cell, 1], \"_\"))[2]\n",
    "metadata[cell, 4] <- gsub(\"\\\\.[0-9]\", \"\", metadata[cell, 4])\n",
    "}\n",
    "```\n",
    "\n",
    "### Metadata File Verification\n",
    "We can verify the accuracy of the file by viewing it `View(metadata)`, and it should look like this:\n",
    "```\n",
    "       sample                        path sex tissue\n",
    "1 female_NF.0 output/female_NF.0.kallisto   F     NF\n",
    "2 female_NF.1 output/female_NF.1.kallisto   F     NF\n",
    "3 female_NF.2 output/female_NF.2.kallisto   F     NF\n",
    "4 female_NP.0 output/female_NP.0.kallisto   F     NP\n",
    "5 female_NP.1 output/female_NP.1.kallisto   F     NP\n",
    "6 female_NP.2 output/female_NP.2.kallisto   F     NP\n",
    "```\n",
    "\n",
    "We can also run the following command to compare it to the metadata included with our downloaded data. Ideally, the output should be a 4x24 matrix containing TRUE in each index. \n",
    "\n",
    "```R\n",
    "default_metadata <- read.table(\"default_metadata.txt\", header = TRUE)\n",
    "metadata == default_metadata\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75cb1c59",
   "metadata": {},
   "source": [
    "### Creating a Data Frame for Gumshoe\n",
    "All the steps we have done up till now are ubiquitous when running a Sleuth analysis. Typically, we would now begin to construct a `sleuth_object`. According to the Sleuth tutorial ([found here](https://pachterlab.github.io/sleuth_walkthroughs/trapnell/analysis.html)), a `sleuth_object` will hold some experimental information, model details for differential testing, and the results from the tests. Although, the method employed by Sleuth to construct this object is very fixed and requires you to run `sleuth_prep` and `sleuth_fit` on all provided metadata files and the associated models, which can require extensive time to code. It's particularly time-consuming and repetitive when, for example, using different models on the same metadata file or comparing results between an analysis using a metadata file that includes a sample or omits it.\n",
    "\n",
    "Gumshoe uses the `sleuth_interpret` function to compensate for this shortcoming; however, particular attention to detail is needed when initially setting up the required data structure, but afterward, the entire process is, for the most part, automated and hands-off. These steps might seem redundant and unnecessary, but let's consider that we wanted to convert the tissue column in our metadata file from a numeric to a factor and compare the analysis between two baseline levels like NF and TH. A typical approach would require double the amount of lines, or we can add this information to our data frame and leave the remainder of the code untouched.\n",
    "\n",
    "First, let's convert the tissue column to a factor and create a new variable to hold our re-leveled metadata.\n",
    "```R\n",
    "# Convert the tissue type column to a factor, and based on how R works, NF will be the first-factor level\n",
    "metadata$tissue <- as.factor(metadata$tissue)\n",
    "\n",
    "# Create a new variable that will contain a re-levelled version of the metadata\n",
    "metadata_releveled <- metadata\n",
    "metadata_releveled$tissue <- relevel(metadata_releveled$tissue, ref = \"TH\")\n",
    "```\n",
    "\n",
    "\n",
    "We can now create a data frame that contains the information that Gumshoe will use to automate the analysis.\n",
    "```R\n",
    "# Define the metadata file name(s), the associated model name(s), and the corresponding model(s) to be used. \n",
    "# The `I` before the mathematical expressions informs R that the expression should remain uninterpreted until \n",
    "# Sleuth processes the model. \n",
    "# NOTE: Model data order must align with the model names\n",
    "metadata_names <- c(\"metadata_nominal_NF_baseline\",\n",
    "                    \"metadata_nominal_TH_baseline\")\n",
    "model_names <- c(\"NF_model_sex,NF_model_sex_tissue,NF_model_interaction\",\n",
    "                 \"TH_model_sex,TH_model_sex_tissue,TH_model_interaction\")\n",
    "model_data <- c(\"~sex, ~sex + tissue, ~sex*tissue\",\n",
    "                \"~sex, ~sex + tissue, ~sex*tissue\")\n",
    "\n",
    "# Take the metadata names, model names, and model data information and combine it into a single data frame\n",
    "analysis_data <- data.frame(metadata_name = metadata_names, metadata_file = tibble(list(metadata)), model_name = model_names, model_data = model_data)\n",
    "```\n",
    "***NOTE: The model names must be unique, or they will be overwritten***\n",
    "\n",
    "Let's go over what we did; we defined two metadata names corresponding to two different metadata files, which have a varying factor baseline, named three models to use for each metadata, and created formulae associated with each model name. Each model uses different factors to run the analysis. The first model is solely sex, the second is sex and tissue type, and the third is sex, tissue type, and the interaction between both.\n",
    "\n",
    "As some of you might have noticed, the `analysis_data` data frame contains identical metadata files for each metadata name. We will need to change the metadata file for the 'metadata_nominal_TH_baseline' from the default file to the intended altered file for our analysis to function, and, to do so, we can use the following command:\n",
    "```R\n",
    "analysis_data[[2]][[2]] <- metadata_releveled\n",
    "```\n",
    "\n",
    "To confirm that we have the correct metadata in the `analysis_data` data frame, we can check the factor levels of the tissue column in the metadata files contained in the  `analysis_data` data frame.\n",
    "```R\n",
    "levels(analysis_data[[2]][[1]]$tissue)\n",
    "# \"NF\"  \"NP\"  \"PEP\" \"TH\" \n",
    "levels(analysis_data[[2]][[2]]$tissue)\n",
    "# \"TH\"  \"NF\"  \"NP\"  \"PEP\"\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c27a3f3",
   "metadata": {},
   "source": [
    "### Using Gumshoe\n",
    "Finally, with a few lines of code, we can analyze all the data. First, we will be using the `sleuth_interpret` function that comes with Gumshoe. We do not need to worry about assigning the function output to a variable, as this step is also automated for you. \n",
    "\n",
    "```R\n",
    "sleuth_interpret(analysis_data)\n",
    "```\n",
    "\n",
    "Gumshoe follows this notion of automation throughout all of its functions. For example, after running `sleuth_interpret`, which runs `sleuth_prep` and `sleuth_fit`, you might wish to run either the wald test (WT) or the likelihood-ratio test (LRT) on all `sleuth_objects` and to accomplish this Gumshoe includes `sleuth_test_wt` and `sleuth_test_lrt`. The test statistics are ran for each coefficent in the models of the `sleuth_object` the code was run on. Similarly, the output of the functions is assigned to the `sleuth_object` the code was run on.\n",
    "```R\n",
    "# Run a Wald test on each sleuth_object\n",
    "sleuth_test_wt(so_NF_model_sex)\n",
    "sleuth_test_wt(so_NF_model_sex_tissue)\n",
    "sleuth_test_wt(so_NF_model_interaction)\n",
    "sleuth_test_wt(so_TH_model_sex)\n",
    "sleuth_test_wt(so_TH_model_sex_tissue)\n",
    "sleuth_test_wt(so_TH_model_interaction)\n",
    "```\n",
    "\n",
    "To retrieve the significant results (FDR < .01) from the `sleuth_object` we can run `sleuth_object_result`, if you require the non-significant results, the `all_data = FALSE` parameter can be excluded. No variable assignment is required at this stage either as Gumshoe automatically assigns the results to a variable.\n",
    "```R\n",
    "# Obtain the results from the sleuth_object\n",
    "sleuth_object_result(so_NF_model_sex, all_data = FALSE, q_max = 0.01)\n",
    "sleuth_object_result(so_NF_model_sex_tissue, all_data = FALSE, q_max = 0.01)\n",
    "sleuth_object_result(so_NF_model_interaction, all_data = FALSE, q_max = 0.01)\n",
    "sleuth_object_result(so_TH_model_sex, all_data = FALSE, q_max = 0.01)\n",
    "sleuth_object_result(so_TH_model_sex_tissue, all_data = FALSE, q_max = 0.01)\n",
    "sleuth_object_result(so_TH_model_interaction, all_data = FALSE, q_max = 0.01)\n",
    "```\n",
    "\n",
    "We should now have all the significant results (FDR < .01) from each `sleuth_object`. At this point, you can save the data or manipulate it as you wish. To save the results we will need to write each result to our drive:\n",
    "```R\n",
    "# Save all the results.\n",
    "# Save the NF baseline results as .txt in the current directory\n",
    "write.table(sig_wald_NF_model_interaction_sexM, row.names = FALSE, sep = \"\\t\", quote = FALSE, \"NF_interaction_model_sexM.txt\")\n",
    "write.table(sig_wald_NF_model_interaction_sexM_tissueNP, row.names = FALSE, sep = \"\\t\", quote = FALSE, \"NF_interaction_model_sexM_NP.txt\")\n",
    "write.table(sig_wald_NF_model_interaction_sexM_tissuePEP, row.names = FALSE, sep = \"\\t\", quote = FALSE, \"NF_interaction_model_sexM_PEP.txt\")\n",
    "write.table(sig_wald_NF_model_interaction_sexM_tissueTH, row.names = FALSE, sep = \"\\t\", quote = FALSE, \"NF_interaction_model_sexM_TH.txt\")\n",
    "write.table(sig_wald_NF_model_interaction_tissueNP, row.names = FALSE, sep = \"\\t\", quote = FALSE, \"NF_interaction_model_NP.txt\")\n",
    "write.table(sig_wald_NF_model_interaction_tissuePEP, row.names = FALSE, sep = \"\\t\", quote = FALSE, \"NF_interaction_model_PEP.txt\")\n",
    "write.table(sig_wald_NF_model_interaction_tissueTH, row.names = FALSE, sep = \"\\t\", quote = FALSE, \"NF_interaction_model_TH.txt\")\n",
    "write.table(sig_wald_NF_model_sex_sexM, row.names = FALSE, sep = \"\\t\", quote = FALSE, \"NF_sex_model_sexM.txt\")\n",
    "write.table(sig_wald_NF_model_sex_tissue_sexM, row.names = FALSE, sep = \"\\t\", quote = FALSE, \"NF_sex_tissue_model_sexM.txt\")\n",
    "write.table(sig_wald_NF_model_sex_tissue_tissueNP, row.names = FALSE, sep = \"\\t\", quote = FALSE, \"NF_sex_tissue_model_NP.txt\")\n",
    "write.table(sig_wald_NF_model_sex_tissue_tissuePEP, row.names = FALSE, sep = \"\\t\", quote = FALSE, \"NF_sex_tissue_model_PEP.txt\")\n",
    "write.table(sig_wald_NF_model_sex_tissue_tissueTH, row.names = FALSE, sep = \"\\t\", quote = FALSE, \"NF_sex_tissue_model_TH.txt\")\n",
    "\n",
    "# Save the TH baseline results as .txt in the current directory\n",
    "write.table(sig_wald_TH_model_interaction_sexM, row.names = FALSE, sep = \"\\t\", quote = FALSE, \"TH_interaction_model_sexM.txt\")\n",
    "write.table(sig_wald_TH_model_interaction_sexM_tissueNF, row.names = FALSE, sep = \"\\t\", quote = FALSE, \"TH_interaction_model_sexM_NF.txt\")\n",
    "write.table(sig_wald_TH_model_interaction_sexM_tissueNP, row.names = FALSE, sep = \"\\t\", quote = FALSE, \"TH_interaction_model_sexM_NP.txt\")\n",
    "write.table(sig_wald_TH_model_interaction_sexM_tissuePEP, row.names = FALSE, sep = \"\\t\", quote = FALSE, \"TH_interaction_model_sexM_PEP.txt\")\n",
    "write.table(sig_wald_TH_model_interaction_tissueNF, row.names = FALSE, sep = \"\\t\", quote = FALSE, \"TH_interaction_model_NF.txt\")\n",
    "write.table(sig_wald_TH_model_interaction_tissueNP, row.names = FALSE, sep = \"\\t\", quote = FALSE, \"TH_interaction_model_NP.txt\")\n",
    "write.table(sig_wald_TH_model_interaction_tissuePEP, row.names = FALSE, sep = \"\\t\", quote = FALSE, \"TH_interaction_model_PEP.txt\")\n",
    "write.table(sig_wald_TH_model_sex_sexM, row.names = FALSE, sep = \"\\t\", quote = FALSE, \"TH_sex_model_sexM.txt\")\n",
    "write.table(sig_wald_TH_model_sex_tissue_sexM, row.names = FALSE, sep = \"\\t\", quote = FALSE, \"TH_sex_tissue_model_sexM.txt\")\n",
    "write.table(sig_wald_TH_model_sex_tissue_tissueNF, row.names = FALSE, sep = \"\\t\", quote = FALSE, \"TH_sex_tissue_model_NF.txt\")\n",
    "write.table(sig_wald_TH_model_sex_tissue_tissueNP, row.names = FALSE, sep = \"\\t\", quote = FALSE, \"TH_sex_tissue_model_NP.txt\")\n",
    "write.table(sig_wald_TH_model_sex_tissue_tissuePEP, row.names = FALSE, sep = \"\\t\", quote = FALSE, \"TH_sex_tissue_model_PEP.txt\")\n",
    "```"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "a5ff606f",
   "metadata": {},
   "source": [
    "### Gumshoe Ensembl to ID and Plotting\n",
    "Gumshoe does include a few other additional functions. This part of the tutorial will briefly cover a few of them like how to convert the Ensembl IDs to gene names using the included `ensembl_to_id` function and creating a volcano plot. We will go over how to annotate a single set of results and generate a volcano plot for them, but the process is the same for all results. The downloadable data contains all the annotated results and associated volcano plots for comparison.\n",
    "\n",
    "First we must create a `mart` object that contains all the information required to retrieve the data needed to annotate our results.\n",
    "```R\n",
    "# Create a mart object to retrive the required information\n",
    "mart <- useMart(\"ensembl\", dataset = \"mmusculus_gene_ensembl\")\n",
    "```\n",
    "\n",
    "The next step is quite simple, we can call the `ensembl_to_id` function on one of the data frames that we have and it will convert the ensembl id's to gene id's. The function automatically assigns the results to a variable called `annotated_sig_results`.\n",
    "```R\n",
    "# Call the ensembl_to_id function on a data frame containing results from the NF baseline metadata file with\n",
    "# the model ~sex\n",
    "ensembl_to_id(sig_wald_NF_model_sex_sexM)\n",
    "```\n",
    "\n",
    "We can now take the `annotated_sig_results` variable and call the `volc_plot` function on it to create a volcano plot of the differentially expressed genes. The `ensembl_to_id` and/or `volc_plot` function can also be called on the non-significant results.\n",
    "```R\n",
    "# Create a volcano plot of the annotated data\n",
    "volc_plot(annotated_sig_results, graph_name = \"Sample Graph - NF Baseline, ~sex, sexM\")\n",
    "```\n",
    "\n",
    "You can find this figure in the downloadable data.\n",
    "\n",
    "***NOTE: If there is a lot of data, the plotting function will take a long time to process.***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9eacf86c",
   "metadata": {},
   "source": [
    "### Conclusion\n",
    "This takes us to the end of the tutorial, and based on our dataset, we have covered nearly all applicable functions. For reference, Gumshoe also includes the following functions that were not used in this tutorial: `design_filter`, `fdr_cutoff` (built-into `sleuth_object_result`), `sleuth_prep_transcript_dosage`, `sleuth_reliable_target_ids`, and `sleuth_u`. Gumshoe also includes several other functions, but they have not been exported and are not available for use right now. If you have any questions, suggestions, or feedback feel free to send an email!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "198eb2ad",
   "metadata": {},
   "source": [
    "---\n",
    "# Supplemental Information\n",
    "\n",
    "### Obtaining The Data and Converting From Single-Cell to Bulk.\n",
    "If you would like to obtain the data yourself, run the following code:\n",
    "\n",
    "***TODO: Add the command that Paul used***\n",
    "\n",
    "### Kallisto Quantification\n",
    "If you would like to quantify the transcript abundances yourself, the following command was used when running Kallisto:\n",
    "\n",
    "***Note the use of `file_1` after the `-o` and in the FASTQ input file portion of the command (More information can be found [here](https://pachterlab.github.io/kallisto/manual))***\n",
    "```R\n",
    "kallisto quant -t 12 -b 40 -i ~/m39.refMrna --single -l 180 -s 20 -o ~/output/file_1.kallisto ~/raw_read/file_1.fastq\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
